equinix/provider.go

```diff
@@ -44,18 +45,25 @@ func Provider() *schema.Provider {
 				Description:  "The Equinix API base URL to point out desired environment. Defaults to https://api.equinix.com",
 			},
 			"client_id": {
-				Type:         schema.TypeString,
-				Optional:     true,
-				DefaultFunc:  schema.EnvDefaultFunc(clientIDEnvVar, nil),
-				ValidateFunc: validation.StringIsNotEmpty,
-				Description:  "API Consumer Key available under My Apps section in developer portal",
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc(clientIDEnvVar, nil),
+				// ValidateFunc: validation.StringIsNotEmpty,
+				Description: "API Consumer Key available under My Apps section in developer portal",
 			},
 			"client_secret": {
-				Type:         schema.TypeString,
-				Optional:     true,
-				DefaultFunc:  schema.EnvDefaultFunc(clientSecretEnvVar, nil),
-				ValidateFunc: validation.StringIsNotEmpty,
-				Description:  "API Consumer secret available under My Apps section in developer portal",
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc(clientSecretEnvVar, nil),
+				// ValidateFunc: validation.StringIsNotEmpty,
+				Description: "API Consumer secret available under My Apps section in developer portal",
+			},
+			"token": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc(clientTokenEnvVar, nil),
+				// ValidateFunc: validation.StringIsNotEmpty,
```

Comment:
I would add conflict checking to avoid confusion over which one takes precedence if both (client-pair/token) are set:

`ConflictsWith: []string{"client_id", "client_secret"},
`

Reply:
ConflictsWith makes sense in the scope of a provider block:
```hcl
provider "equinix" {
   client_id ="..."
   client_secret = "..."
   token = "..."
}
```

But we may also want to consider users supplying secrets via environment.
I'm not sure if we should reject `EQUINIX_TOKEN` if `EQUINIX_CLIENT_ID` or `EQUINIX_CLIENT_SECRET` are set.

Perhaps `ConflictsWith` will not restrict our ability to handle environment variables.
I'd like to compare notes with other Terraform providers to see what they've done in this situation.

Reply:
As you mention, my proposal could be a problem with environment variables and with external credential files (possible alternative in the future).

After checking the official providers, none of them add restrictions at schema definition, and all conditions are managed in the Config builder function ([ex GCP Provider](https://github.com/hashicorp/terraform-provider-google/blob/f98a7c02b43c4113253a547a2d1ad213156a45a1/google/config.go#L1101)) or underlying sdk. However, they do define the order of precedence in the attribute [documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#:~:text=application%2Ddefault%20login.-,access_token,-%2D%20(Optional)%20A%20temporary)

> If you specify both with environment variables, Terraform uses the access_token instead of the credentials field

Reply:
I've updated the provider documentation, I think this takes care of both concerns.